apps/axistream/src/axi_stream_reader.hh

```diff
@@ -0,0 +1,168 @@
+#pragma once
+
+#include <type_traits>
+#include "hls_macros.hh"
+#include "util.hh"
+#include "axi_stream_common.hh"
+#include "message.hh"
+
+/*
+ * Reads arbitrary length data (in bytes) from an AXI stream.
+ * This adheres to the Xilinx AXI Reference Guide UG 761:
+ * https://www.xilinx.com/support/documentation/ip_documentation/ug761_axi_reference_guide.pdf
+ *
+ * This standard differs from the ARM standard
+ * (https://static.docs.arm.com/ihi0051/a/IHI0051A_amba4_axi4_stream_v1_0_protocol_spec.pdf)
+ * in the following key ways:
+ * TKEEP: is ignored, since null bytes are only used in a packet with TLAST set, and the user of this API requests a packet by length.
+ * TID, TDEST, TUSER: are all currently ignored, even though Xilinx supports them. Note that the ap_axiu type requires a bit width of at
+ * least 1 for all of these signals, even though they are unused, in accordance with UG 761.
+ * (Also, the tools won't accept 0 bit signals in Verilog anyway, but will delete unused 1 bit signals).
+ */
+template <int DataWidth>
+class AxiStreamReader {
+public:
+    AxiStreamReader(AxiStream<DataWidth>& stream) : stream_(stream) {
```

Comment:
Do you think it is possible to make this a drop-in replacement for the existing classes?

Reply:
Yes. It roughly is now; the only difference are the method names (`get` vs `read, etc): https://github.com/Xilinx/HLS_packet_processing/blob/master/apps/common/ip.hpp#L2811

Reply:
Would it make sense to modify the existing class to take a message as another option?  This could make it easier to do an apples-to-apples comparison.

Reply:
Then we could just see the difference of the parsing code. It's probably possible to do that (at least in some hacky temporary way), the key is this `set_le` call: https://github.com/Xilinx/HLS_packet_processing/blob/master/apps/common/ip.hpp#L2869. It also expects a `LENGTH` member: https://github.com/Xilinx/HLS_packet_processing/blob/master/apps/common/ip.hpp#L2829. But you could make a message conform to those requirements, and the `LittleEndianByteReader` should work I think.

It might be easier to try both systems on say an Ethernet + ipv4 header independently and just see the difference.

Reply:
I think both are worthwhile to understand the differences.  I'm loathe to give up proper handling of KEEP/LAST.  I like your message abstraction and I think it could make the code simpler for defining packet formats.

Reply:
I have a version which handles keep/last correctly (prevents you from running off the end) that I'm experimenting with; once I get it working I'll add it.

Reply:
I'm trying to make it parameterizable on how it handles length checking, because once you do keep/last correctly, the loop exit condition (or the if inside the loop) is data-dependent, which seems to increase the iteration latency of the loop, but I'm still experimenting.

Reply:
I think that bit is well handled in the original code.  I suspect you're just going to end up recreating what is there.  I'd like to suggest focusing on adapting those classes to support your message type.